### PR TITLE
wf-recorder --help / manpage

### PR DESCRIPTION
**Pull request by Jpfonseca**
Started to contribute to fulfill issue #26

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -507,13 +507,14 @@ int main(int argc, char *argv[])
...
         { "to-yuv",          no_argument,       NULL, 't' },
+        { "help",            no_argument,       NULL, 'h' },
         { 0,                 0,                 NULL,  0  }
     };
 
     int c, i;
     std::string param;
     size_t pos;
-    while((c = getopt_long(argc, argv, "o:f:g:c:p:d:la::t::", opts, &i)) != -1)
+    while((c = getopt_long(argc, argv, "o:f:g:c:p:d:la::t::h::", opts, &i)) != -1)
```
Comment (soreau) [inline src/main.cpp]: From the getopt_long man page `Two colons mean an option takes an optional arg`. This option doesn't take an optional argument, does it?

File: src/main.cpp [hunk 2]
```diff
@@ -563,12 +565,65 @@ int main(int argc, char *argv[])
                     printf("Invalid codec option %s\n", optarg);
                 }
                 break;
-
+
```
Comment (soreau) [inline src/main.cpp]: Don't add random whitespace.

File: src/main.cpp [hunk 3]
```diff
@@ -563,12 +565,65 @@ int main(int argc, char *argv[])
...
                 break;
-
+            
             default:
                 printf("Unsupported command line argument %s\n", optarg);
         }
     }
-
+    
+    if(c=='h'){
```
Comment (soreau) [inline src/main.cpp]: Use spaces around `==` and after `)`.

File: src/main.cpp [hunk 4]
```diff
@@ -563,12 +565,65 @@ int main(int argc, char *argv[])
...
+         printf("Usage: wf-recorder [OPTION]... [FILE]...
```
Comment (soreau) [inline src/main.cpp]: This printf statement is not aligned with the rest of the code.

File: src/main.cpp [hunk 5]
```diff
@@ -563,12 +565,65 @@ int main(int argc, char *argv[])
...
+  -g, --geometry            Selects a specific part of the screen.
+  
+  -h, --help                Prints this help screen.
+                
+  -l, --log                 Generates a log on the current terminal. Debug purposes. 
+                
+  -o, --output              Specify the output where the video is to be recorded.
+                
+  -p, --codec-param         Change the codec parameters.
+                       -p <option_name>=<option_value>
```
Comment (soreau) [inline src/main.cpp]: The alignment is incorrect here too. Use spaces instead of tabs.

File: src/main.cpp (hunk 1 above)
Comment (Jpfonseca) [reply src/main.cpp]: Sure misclick on that. Btw can I change the options to be in alphabetical order ?

**Comment by Jpfonseca** [comment]
Made the changes needed. Can I change the options to be in alphabetical order ?

File: src/main.cpp [hunk 6]
```diff
@@ -549,7 +550,10 @@ int main(int argc, char *argv[])
             case 't':
                 params.to_yuv = true;
                 break;
-
+
```
Comment (soreau) [inline src/main.cpp]: This line adds whitespace, several spaces on a blank line.

File: src/main.cpp [hunk 7]
```diff
@@ -569,6 +573,59 @@ int main(int argc, char *argv[])
         }
     }
 
+    if ( c == 'h' ) {
```
Comment (soreau) [inline src/main.cpp]: Don't put space after `(` or before `)`.

**Comment by soreau** [comment]
> Can I change the options to be in alphabetical order ?

I think this is a different idea that should be in a separate PR. @ammen99 might have different ideas than me, what I say is not absolute.

**Comment by soreau** [comment]
> @soreau I just looked to the rest of the file and changed the next if to follow the same code style also

I notice the file is quite inconsistent in this regard, the brackets should all be the same style though I don't know which way is correct. Really need style guidelines somewhere. @ammen99?

**Comment by ammen99** [comment]
I'm on a trip so I can't review properly atm, I'll be back in 3 days. Anyways, brackets should be on the next line.

**Comment by Jpfonseca** [comment]
Will keep that in mind.

File: src/main.cpp [hunk 8]
```diff
@@ -569,8 +573,63 @@ int main(int argc, char *argv[])
         }
     }
 
+    if (c == 'h')
```
Comment (ammen99) [inline src/main.cpp]: Please move this to a separate function.

File: src/main.cpp [hunk 9]
```diff
@@ -569,8 +573,63 @@ int main(int argc, char *argv[])
         }
     }
 
+    if (c == 'h') 
+    {
+         printf("Usage: wf-recorder [OPTION]... [FILE]...
```
Comment (ammen99) [inline src/main.cpp]: `[FILE]` is passed with the `-f` parameter, not standalone.

File: src/main.cpp [hunk 10]
```diff
@@ -569,8 +573,63 @@ int main(int argc, char *argv[])
...
+                
+                
+Examples:
+  
+  Video Only:
+  
+  - wf-recorder                       Records the video. Use Ctrl+C to stop recording. 
+                                      The video file will be stored as recording.mp4 in the current working directory.
+                
+  - wf-recorder -f <filename>         Records the video. Use Ctrl+C to stop recording.
```
Comment (ammen99) [inline src/main.cpp]: The output name can be another one. For example, the user can also set `wf-recorder -f test.mkv` and then the resulting file will be `test.mkv` not `test.mp4`

File: src/main.cpp [hunk 11]
```diff
@@ -569,8 +573,63 @@ int main(int argc, char *argv[])
...
     }
 
+    if (c == 'h') 
+    {
+         printf("Usage: wf-recorder [OPTION]... [FILE]...
+Screen recording of wlroots-based compositors
+
+With no FILE, start recording the current screen.
+
+  -a, --audio               Starts recording the screen with audio
```
Comment (ammen99) [inline src/main.cpp]: For the sake of completeness, the `--audio/-a` parameter takes an optional parameter, which is the name of the pulseaudio device to capture (the list can be obtained by running `pactl list sinks | grep Name`)

File: src/main.cpp [hunk 12]
```diff
@@ -569,8 +573,63 @@ int main(int argc, char *argv[])
...
+  -h, --help                Prints this help screen.
+                
+  -l, --log                 Generates a log on the current terminal. Debug purposes. 
+                
+  -o, --output              Specify the output where the video is to be recorded.
+                
+  -p, --codec-param         Change the codec parameters.
+                            -p <option_name>=<option_value>
+
+  -t, to-yuv                Use the -t or --to-yuv option in addition to the vaapi options to convert the data in software,
```
Comment (ammen99) [inline src/main.cpp]: Maybe leave this out from this PR until we merge #32, then me or @soreau will add the docs for `-t` and `-e`

File: src/main.cpp (hunk 10 above)
Comment (Jpfonseca) [reply src/main.cpp]: On the first example no arguments where given. And for what I have experienced it always results in the test.mp4. Probably I can add the information of the inexistence of  arguments

Comment (ammen99) [reply src/main.cpp]: It is just that the default file name is `recording.mp4`. If you specify a `-f test.mkv` then it should produce `test.mkv`. If it doesn't, then this is a bug.

Comment (Jpfonseca) [reply src/main.cpp]: Let me clarify, I am giving 2 different examples on this print.

The first with the default value of the command `wf-recorder` without any arguments.
On the  the second, with a specific filename by running the command `wf-recorder -f <filename>` which will result in a recording with the name `filename`

Comment (soreau) [reply src/main.cpp]: Perhaps you could elaborate the comment for the -f case (or both comments if necessary) to explain that `-f file.ext` means that the output file will be called `file.ext` and the file format will be guessed by the file extension `.ext`, which will fail if it's not one listed in the output of `ffmpeg -muxers`.

Comment (ammen99) [reply src/main.cpp]: The whole point is that the output file is not necessarily a .mp4 file, as the current help message says(and as far as I understand it)

File: src/main.cpp (hunk 12 above)
Comment (soreau) [reply src/main.cpp]: @ammen99 #30 was merged so -t already exists.

File: src/main.cpp (hunk 10 above)
Comment (Jpfonseca) [reply src/main.cpp]: @ammen99  Sorry I only saw the .mp4 on the second example now, as I was answering on my phone :/ .  You were right  from the beginning. 
@soreau  I will try to address that.

File: src/main.cpp (hunk 12 above)
Comment (Jpfonseca) [reply src/main.cpp]: Didn't add the -e option

**Comment by Jpfonseca** [comment]
> I think it's safe to omit `Supports GIF output also.`. You could mention the list can be found with `ffmpeg -encoders`, though.

GIF support was the main reason why I started using this tool. If I hadn't read an issue on it, I wouldn't be using it

**Comment by Jpfonseca** [comment]
@soreau  just tested everything and made the changes you asked

File: src/main.cpp [hunk 13]
```diff
@@ -481,6 +481,74 @@ static wf_recorder_output* detect_output_from_region(const capture_region& regio
     return nullptr;
 }
 
+static void help(){
```
Comment (ammen99) [inline src/main.cpp]: new line :)

File: src/main.cpp [hunk 14]
```diff
@@ -481,6 +481,74 @@ static wf_recorder_output* detect_output_from_region(const capture_region& regio
...
+
+    printf("\
```
Comment (ammen99) [inline src/main.cpp]: You can use C++11 raw strings here. Just do

```
printf(R"(
blah
blah
)");

```

In this case, you won't have to add `\n\` to every line.

Comment (Jpfonseca) [reply src/main.cpp]: Will change it .  Didn't knew about this improvement. Also was took too much time doing the `\n\ ` way...

**Comment by Jpfonseca** [comment]
Applied all the changes needed. 
@ammen99 any ideas on what to use to make the man page ?  `scdoc` ?
That probably the best way to keep the manpage simple stupid whilst not messing much with the `meson.build` file .

**Comment by ammen99** [comment]
I'd use whatever sway uses(which I think is scdoc)

**Comment by Jpfonseca** [comment]
Added first version of the manpage.
@ammen99  @soreau do you think I should change anything else ? Or add something ?

**Comment by Jpfonseca** [comment]
Was a good ride. At least I now know how to make a manpage 🎉.

**Pull request merged**